File: js/algorithm/bin_search.js
/**
 * 二分查找法，只能针对有序的数组进行查找， 折半查找法
 */

function binSearch(list, target) {
    let upperBound = list.length - 1;
    let lowerBound = 0;
    // 当查找区间中没有该元素时，返回-1
    // if (target > list[list.length - 1] || target < list[0]) {
    //     return -1
    // }
    while (lowerBound <= upperBound) {
        const mid = Math.floor((lowerBound + upperBound) / 2)
        if (list[mid] < target) {
            lowerBound = mid + 1
        } else if (list[mid] > target) {
            upperBound = mid - 1
        } else {
            return mid
        }
    }
    return -1
}

module.exports = binSearch

File: py/algorithm/bin_search.py
# 二分法查找
from math import floor


# 只适合有序的数组
def bin_search(search_list, target):
    upper = len(search_list) - 1
    lowwer = 0
    # search_list.sort()
    while lowwer <= upper:
        mid_len = floor((upper + lowwer) / 2)
        cur_item = search_list[mid_len]
        # print(mid_len)

        if cur_item < target:
            lowwer = mid_len + 1
        elif cur_item > target:
            upper = mid_len - 1
        else:
            return mid_len
    return -1
